### DISCUSSION TITLE
Implementing reboot to bootloader in QT

### DISCUSSION THREAD

**Discussion by SeongGino** [earlephilhower/arduino-pico#2119] [Q&A]
Hi! A quick question regarding the magic number reboot:

I'm developing an app to communicate with RP2040 microcontrollers (using a specific firmware) over serial. For what it's worth, every back-and-forth communication between the app and the board works as I'd programmed it, but for the sake of convenience I'd like to be able to signal the board to reboot so I can have the app automagically copy an updated firmware over.

I have tried using the relevant function from `tools/uf2conv.py`:
https://github.com/earlephilhower/arduino-pico/blob/d53d0033fe1870df29714348f166d94118a0f9fc/tools/uf2conv.py#L374-L382
...And essentially re-implemented it with equivalent settings of QSerialPort:
```
// NOTE: serial port number/path is already set & at 9600bps before this.
    serialPort.setDataTerminalReady(true);
    QThread::msleep(100);
    serialPort.setDataTerminalReady(false);
    serialPort.setBaudRate(QSerialPort::Baud1200);
    serialPort.close();
```
However, even with if checks to ensure the port is using the set 1200bps passing, the board doesn't reboot--it's still open. I've tried sending a single "." after the serial port baud rate setting command, but still no difference. (For what it's worth, normal app-board communications is done over 9600bps, and would be open at the time the button to trigger this function is pressed.)

Since I'm on Linux, if I use `stty -F /dev/ttyACM# 1200` (where # is the port number, usually 0 since I'm testing with one board at a time), the board immediately reboots to bootloader. So I'm wondering what the difference is, and if there's a solution - anyone else better versed in QT to respond would be appreciated.

(While I probably COULD have the app just run that stty command itself, unfortunately this app is made to be cross-platform with Windows, and that OS does not have as simple a method to communicate with COM devices in a single cmd)

**Comment by earlephilhower** [comment]
You need to set the virtual DTR low as well as be at 1200BPS.  See [the code](https://github.com/earlephilhower/arduino-pico/blob/d53d0033fe1870df29714348f166d94118a0f9fc/cores/rp2040/SerialUSB.cpp#L183).  Python and `stty` may do this automatically, but in Qt maybe you need to specify it manually?

**Reply by SeongGino** [reply]
Thank you very much for the reply, that snippet definitely clarified things a bit.

Unfortunately, even when setting DTR low (which has to be done while the port is open), closing the port, setting 1200bps, and re-opening the port, I'm still not getting any such result (despite my debug messages telling me that these settings were indeed set). It's very possible Qt's implementation might be a little off, so I'm likely to follow up with them later - it wouldn't be the first time I've had inconsistent results with Qt's Serial Port.

**Reply by earlephilhower** [reply]
Maybe try setting 9600, sending `NUL`, then setting 1200, then setting an explicit DTR=1, send a dummy `NUL`, then DTR=0 and send another dummy `NUL`?  This is all USB pretending to be a serial port so maybe the transition isn't actually being packetized and sent down the wire unless there's data?
